File: go/toolset.go
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

package copilot

import (
	"fmt"
	"regexp"
)

// ClientMode controls the default surface presented to sessions created by the
// [Client]. The zero value is [ModeCopilotCli], matching the legacy CLI defaults.
//
// Set [ClientOptions.Mode] to [ModeEmpty] to opt in to multi-tenant safe
// defaults: no built-in tools by default (callers must specify
// [SessionConfig.AvailableTools] explicitly), no environment_context section
// in the system message, telemetry off, custom instructions and remote-custom
// agents disabled, etc.
type ClientMode string

const (
	// ModeCopilotCli is the default mode; sessions inherit the full Copilot
	// CLI experience (all built-in tools, host environment_context, etc.).
	ModeCopilotCli ClientMode = "copilot-cli"
	// ModeEmpty is the multi-tenant safe-default mode. Sessions start with
	// no built-in tools, no environment context, and various features
	// (custom instructions, remote agents, telemetry, plugins) off by
	// default. Callers can opt back in field-by-field.
	ModeEmpty ClientMode = "empty"
)

// ToolSet builds a list of source-qualified tool filter patterns
// (`builtin:*`, `mcp:<name>`, `custom:*`, ...) for use with
// [SessionConfig.AvailableTools] or [SessionConfig.ExcludedTools].
//
// Tools are classified by the runtime at registration time (not from name
// parsing), so AddBuiltIn("foo") matches only tools the runtime registered as
// built-in, even if an MCP server or custom-agent extension happens to
// register a tool with the same wire name.
//
// ToolSet's zero value is ready to use. Convert to []string via [ToolSet.ToSlice]
// before passing to [SessionConfig] fields, e.g.
// `(&ToolSet{}).AddBuiltIn(...).ToSlice()`.
type ToolSet struct {
	items []string
}

// NewToolSet returns an empty [ToolSet].
func NewToolSet() *ToolSet { return &ToolSet{} }

var toolNameRegex = regexp.MustCompile(`^[a-zA-Z0-9_-]+$`)

// AddBuiltIn adds one or more built-in tool patterns. Pass a specific tool
// name (e.g. "bash") or "*" to match all built-in tools.
func (s *ToolSet) AddBuiltIn(names ...string) *ToolSet {
	for _, n := range names {
		validateToolName("builtin", n)
		s.items = append(s.items, "builtin:"+n)
	}
	return s
}

// AddCustom adds a custom-tool pattern. Matches tools registered via
// [SessionConfig.Tools] or via custom agents.
func (s *ToolSet) AddCustom(name string) *ToolSet {
	validateToolName("custom", name)
	s.items = append(s.items, "custom:"+name)
	return s
}

// AddMcp adds an MCP tool pattern. Matches tools advertised by any configured
// MCP server.
func (s *ToolSet) AddMcp(toolName string) *ToolSet {
	validateToolName("mcp", toolName)
	s.items = append(s.items, "mcp:"+toolName)
	return s
}

// ToSlice returns a defensive copy of the accumulated filter strings.
func (s *ToolSet) ToSlice() []string {
	out := make([]string, len(s.items))
	copy(out, s.items)
	return out
}

func validateToolName(kind, name string) {
	if name == "" {
		panic(fmt.Sprintf("invalid %s tool name: must not be empty", kind))
	}
	if name == "*" {
		return
	}
	if !toolNameRegex.MatchString(name) {
		panic(fmt.Sprintf(
			"invalid %s tool name %q: tool names must match /^[a-zA-Z0-9_-]+$/ or be the wildcard %q",
			kind, name, "*"))
	}
}

// BuiltInToolsIsolated lists built-in tools that operate only within the
// bounds of a single session — no host filesystem access outside the session,
// no cross-session state, no host environment access, no network. Safe to
// enable in [ModeEmpty] scenarios (e.g. multi-tenant servers) without leaking
// host capabilities.
//
// Contract: tools in this set MUST NOT be extended (even behind options or
// args) to read or write state outside the session boundary. Adding
// cross-session or host-state behavior to one of these tools is a breaking
// change that requires removing it from this set.
var BuiltInToolsIsolated = []string{
	"ask_user",
	"task_complete",
	"exit_plan_mode",
	"task",
	"read_agent",
	"write_agent",
	"list_agents",
	"send_inbox",
	"context_board",
	"skill",
}

File: java/src/main/java/com/github/copilot/rpc/ToolSet.java
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

package com.github.copilot.rpc;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Objects;
import java.util.regex.Pattern;

/**
 * Builder for {@link SessionConfig#setAvailableTools(java.util.List)} /
 * {@link SessionConfig#setExcludedTools(java.util.List)} using source-qualified
 * filter patterns ({@code builtin:*}, {@code mcp:<name>}, {@code custom:*},
 * etc.).
 * <p>
 * Tools are classified by the runtime at registration time (not from name
 * parsing), so {@link #addBuiltIn(String)} matches only tools the runtime
 * registered as built-in, even if an MCP server or custom-agent extension
 * happens to register a tool with the same wire name.
 * <p>
 * {@code ToolSet} extends {@link ArrayList} so instances can be passed directly
 * to {@link SessionConfig#setAvailableTools(java.util.List)} or
 * {@link SessionConfig#setExcludedTools(java.util.List)}.
 *
 * <h2>Example</h2>
 *
 * <pre>{@code
 * var session = client
 * 		.createSession(new SessionConfig()
 * 				.setAvailableTools(new ToolSet().addBuiltIn(BuiltInTools.ISOLATED).addMcp("*").addCustom("*")))
 * 		.get();
 * }</pre>
 *
 * @since 1.3.0
 */
public class ToolSet extends ArrayList<String> {

    private static final Pattern VALID_TOOL_NAME = Pattern.compile("^[a-zA-Z0-9_-]+$");

    /**
     * Adds a built-in tool pattern.
     *
     * @param name
     *            a specific built-in tool name (e.g. {@code "bash"}) or {@code "*"}
     *            to match all built-in tools
     * @return this {@code ToolSet} for chaining
     * @throws IllegalArgumentException
     *             if name is null, empty, or contains invalid characters
     */
    public ToolSet addBuiltIn(String name) {
        validateName("builtin", name);
        add("builtin:" + name);
        return this;
    }

    /**
     * Adds a list of built-in tool patterns (e.g. {@link BuiltInTools#ISOLATED}).
     *
     * @param names
     *            built-in tool names to add
     * @return this {@code ToolSet} for chaining
     * @throws NullPointerException
     *             if names is null
     */
    public ToolSet addBuiltIn(Collection<String> names) {
        Objects.requireNonNull(names, "names must not be null");
        for (String name : names) {
            addBuiltIn(name);
        }
        return this;
    }

    /**
     * Adds a custom tool pattern. Matches tools registered via the SDK's
     * {@link SessionConfig#setTools(java.util.List)} option or via custom agents.
     *
     * @param name
     *            a specific custom tool name or {@code "*"} to match all custom
     *            tools
     * @return this {@code ToolSet} for chaining
     * @throws IllegalArgumentException
     *             if name is null, empty, or contains invalid characters
     */
    public ToolSet addCustom(String name) {
        validateName("custom", name);
        add("custom:" + name);
        return this;
    }

    /**
     * Adds an MCP tool pattern. Matches tools advertised by any configured MCP
     * server.
     *
     * @param toolName
     *            the runtime's canonical wire name for the MCP tool (e.g.
     *            {@code "github-list_issues"}), or {@code "*"} to match all MCP
     *            tools from any server
     * @return this {@code ToolSet} for chaining
     * @throws IllegalArgumentException
     *             if toolName is null, empty, or contains invalid characters
     */
    public ToolSet addMcp(String toolName) {
        validateName("mcp", toolName);
        add("mcp:" + toolName);
        return this;
    }

    private static void validateName(String kind, String name) {
        if (name == null || name.isEmpty()) {
            throw new IllegalArgumentException("Invalid " + kind + " tool name: must not be null or empty.");
        }
        if ("*".equals(name)) {
            return;
        }
        if (!VALID_TOOL_NAME.matcher(name).matches()) {
            throw new IllegalArgumentException("Invalid " + kind + " tool name '" + name
                    + "': tool names must match /^[a-zA-Z0-9_-]+$/ or be the wildcard '*'.");
        }
    }
}
